Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,799 @@
+product topology, product bornology and product uniformity, to be able to use freely continuity
+statements for the coordinate functions, for instance.
+
+# Implementation notes
+
+This files is a straight-forward adaption of `Mathlib.Analysis.NormedSpace.PiLp`. We deviate from
+`PiLp` in that we use for `p = 0` the junk value `d(x, y) = 0`.
+
+-/
+
+local macro_rules | `($x ^ $y)   => `(HPow.hPow $x $y) -- Porting note: See issue #2220
+
+open Real Set Filter IsROrC Bornology BigOperators Uniformity Topology NNReal ENNReal
+
+noncomputable section
+
+/-- A copy of a Prod type, on which we will put the `L^p` distance. Since the Prod type itself is
+already endowed with the `L^∞` distance, we need the type synonym to avoid confusing typeclass
+resolution. Also, we let it depend on `p`, to get a whole family of type on which we can put
+different distances. -/
+@[nolint unusedArguments]
+def ProdLp (_p : ℝ≥0∞) (α β : Type _) : Type _ :=
+  α × β
+
+instance ProdLp.instInhabited (p : ℝ≥0∞) (α β : Type _) [Inhabited α] [Inhabited β] :
+    Inhabited (ProdLp p α β) :=
+  ⟨default, default⟩
+
+@[ext]
+protected theorem ProdLp.ext {p : ℝ≥0∞} {α β : Type _} {x y : ProdLp p α β}
+    (h1 : x.fst = y.fst) (h2 : x.snd = y.snd) : x = y := Prod.ext h1 h2
+
+namespace ProdLp
+
+variable (p : ℝ≥0∞) (𝕜 𝕜' : Type _) (α β : Type _)
+
+/-- Canonical bijection between `ProdLp p α β` and the original Prod type. We introduce it to be
+able to compare the `L^p` and `L^∞` distances through it. -/
+protected def equiv : ProdLp p α β ≃ α × β :=
+  Equiv.refl _
+
+/-! Note that the unapplied versions of these lemmas are deliberately omitted, as they break
+the use of the type synonym. -/
+
+@[simp]
+theorem equiv_apply (x : ProdLp p α β) : ProdLp.equiv p α β x = x :=
+  rfl
+
+@[simp]
+theorem equiv_symm_apply (x : α × β) : (ProdLp.equiv p α β).symm x = x :=
+  rfl
+
+section DistNorm
+
+/-!
+### Definition of `edist`, `dist` and `norm` on `ProdLp`
+
+In this section we define the `edist`, `dist` and `norm` functions on `ProdLp p α β` without
+assuming `[Fact (1 ≤ p)]` or metric properties of the spaces `α` and `β`. This allows us to provide
+the rewrite lemmas for each of three cases `p = 0`, `p = ∞` and `0 < p.to_real`.
+-/
+
+
+section Edist
+
+variable [EDist α] [EDist β]
+
+/-- Endowing the space `ProdLp p α β` with the `L^p` edistance. We register this instance
+separate from `ProdLp.instPseudoEMetric` since the latter requires the type class hypothesis
+`[Fact (1 ≤ p)]` in order to prove the triangle inequality.
+
+Registering this separately allows for a future emetric-like structure on `ProdLp p α β` for `p < 1`
+satisfying a relaxed triangle inequality. The terminology for this varies throughout the
+literature, but it is sometimes called a *quasi-metric* or *semi-metric*. -/
+instance instEDist : EDist (ProdLp p α β) where
+  edist f g :=
+    if _hp : p = 0 then 0
```

Comment:
Shouldn't this be the count of differing elements for p=0?

Reply:
```suggestion
    if _hp : p = 0 then
        (if f.fst = g.fst then 0 else 1) + (if f.snd = g.snd then 0 else 1)
```

Reply:
I don't know, `PiLp` does this, but everything is basically developed only for `p > 0`. I can change it to match `PiLp`, but I did not find a very nice way to do it.

Reply:
I don't see what you mean by that; all the normed instances are in place for any `p`.